### PR TITLE
apps/btc: move bip143_sighash to Rust 

### PR DESCRIPTION
**Pull request by benma**
Builds on https://github.com/digitalbitbox/bitbox02-firmware/pull/670

### CODE REVIEW

**Review by x1ddos — COMMENTED** [review]
> Builds on #670

What prevents this from compiling with sha2 crate v0.8.1?

**Comment by benma** [comment]
> What prevents this from compiling with sha2 crate v0.8.1?

Incompatible API (they renamed input() to update() and result() to finalize()).

**Comment by x1ddos** [comment]
> > What prevents this from compiling with sha2 crate v0.8.1?
> 
> Incompatible API (they renamed input() to update() and result() to finalize()).

yeah sure. I meant why upgrade to v0.9.2 in #670. If you kept `input` and `result`, and stayed on v0.8.1, would it still work?

**Comment by benma** [comment]
> > > What prevents this from compiling with sha2 crate v0.8.1?
> > 
> > 
> > Incompatible API (they renamed input() to update() and result() to finalize()).
> 
> yeah sure. I meant why upgrade to v0.9.2 in #670. If you kept `input` and `result`, and stayed on v0.8.1, would it still work?

Yeah it would.

**Comment by x1ddos** [comment]
> > > > What prevents this from compiling with sha2 crate v0.8.1?
> > > 
> > > 
> > > Incompatible API (they renamed input() to update() and result() to finalize()).
> > 
> > 
> > yeah sure. I meant why upgrade to v0.9.2 in #670. If you kept `input` and `result`, and stayed on v0.8.1, would it still work?
> 
> Yeah it would.
> 
> Edit: why upgrade: why not? good to be up to date.

Then why not upgrade the others like noise-rust-crypto? It also uses sha2 crate and we now have both, sha2 v0.8.1 and v0.9.2.

**Comment by benma** [comment]
> Then why not upgrade the others like noise-rust-crypto? It also uses sha2 crate and we now have both, sha2 v0.8.1 and v0.9.2.

I didn't realize that. In that case yes I should also upgrade the rest.

**Comment by x1ddos** [comment]
sound good!

File: src/rust/bitbox02-rust/src/apps/bitcoin/bip143.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                *b"cccccccccccccccccccccccccccccccc",
+                54,
+                b"abcdefgh123",
+                85899345922,
+                0xFFFFFFFF - 1,
+                *b"dddddddddddddddddddddddddddddddd",
+                345432345,
+                234234,
+            ),
+            *b"\x24\x09\x34\x00\xb2\x38\x38\x0f\x80\x67\x69\xc2\xe4\xf7\xe2\x18\x6d\x5d\x8e\xd5\xa3\xcc\x6c\x6d\xcd\x64\xf0\x9e\x72\x19\x9b\x71"
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust/src/apps/bitcoin/bip143.rs]: Done

**Comment by benma** [comment]
Updated, PTAL

**Review by x1ddos — APPROVED** [review]
LGTM

File: src/rust/bitbox02-rust-c/src/app_bitcoin.rs [hunk 2]
```diff
@@ -67,3 +69,44 @@ pub unsafe extern "C" fn rust_bitcoin_keypath_validate_address_multisig(
...
+        prevout_index,
+        sighash_script.as_ref(),
+        prevout_value,
+        sequence,
+        core::slice::from_raw_parts(hash_outputs, 32)
+            .try_into()
+            .unwrap(),
+        locktime,
+        sighash_flags,
+    );
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/bitbox02-rust-c/src/app_bitcoin.rs]: Done, in the Rust and FFI layer.

[the remark this answers is not in the document]
Comment (x1ddos) [reply src/rust/bitbox02-rust-c/src/app_bitcoin.rs]: nice. thanks!

**Pull request merged**